### PR TITLE
Add NeuronexusRawIO/IO

### LINKED ISSUE #1296: Support for Neuronexus (.XDAT) files
Hi. I am new to analyzing ephys data, and our lab currently uses allego neuronexus files (.xdat). In wanting to use spike interface, we reached out to see if there was support on their end for .xdat files. 

I wanted to reach out to see if there was developed support for those files or if it could be done.

According to spikeinterface people, when they reached out to Neuronexus they did not seem interested in collaboration. I know that there are features that allow one to include their own file format. If that could be done by me, I can try; however, if the spikeinterface developers are unable to support .xdat file, I definitely don't think I will be able to make that happen.

### PR DESCRIPTION
**Pull request by zm711**
Fixes #1296.

I'll try to work on this a bit in my free time. Will switch off draft when ready for review.

Also should help for [neuroconv](https://github.com/catalystneuro/neuroconv/issues/702)

* [x] Add tests
* [x] Get file on GIN

### CODE REVIEW

**Comment by zm711** [comment]
No rush on this @apdavison, this is a new io. Since Sam will be going on vacation, I think getting some feedback would be nice, but I can also pester him once he is back. There is no rush on this one :)

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [inline neo/rawio/neuronexusrawio.py:50]: I guess the company is technically `NeuroNeuxs` maybe I change that?

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:55]:
```suggestion
    def __init__(self, filename: str | Path):
```
Why add a default value here?

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:69]:
```suggestion
            * The *_timestamps.xdat binary file of the timestamp data
```

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:70]:
```suggestion
        From the metadata, the other two files are located within the same directory
```
This is actually a question to you.

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:102]: is_file also does the function of exists so there is some redundancy here.

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:109]: Nice and informative checks!

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:126]: Dude, where did you get the specification or how did you figured this one out?

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:149]: It would be good to show which ones.

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:159]: I think this should be string to not add more confusing to the already confusing stream_id vs stream_index. At least type increases the gap.

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:275]: Yei!

File: neo/test/iotest/test_neuronexusio.py
Comment (h-mayorquin) [inline neo/test/iotest/test_neuronexusio.py:17]: is this already in gin? Nice : D

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:251]: I like to add path (hungarian notation) to differentiate between file handles (output of f = open() and paths) but I think name works as well!

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [inline neo/rawio/neuronexusrawio.py:27]: What do you mean by this?

**Review by h-mayorquin — APPROVED** [review]
Wow, can I approve things? How comes?  Just testing.

Did the first reading, this looks pretty good to me. A bunch of minor comments.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:55]: It's been like that for ages. Every RawIO has a default. There has been a discussion about switching the default to None, but there is always a default.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:70]: There's not really a hard rule here. Growing up I was taught that 4 words is the dividing line between comma and no comma but part of this is stylistic. Do you want to emphasize the separation of the thoughts or not.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:102]: I think this might be residual from an old implementation that I switched. I think the original idea was someone might get a folder (which could exist) and so I added the check for being a file. But not sure anymore why I put both in. Haha.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:149]: Do you want to do an actual suggestion? You want me to compare the timestamp file from the metadata file in the actual error?

File: neo/test/iotest/test_neuronexusio.py
Comment (zm711) [reply neo/test/iotest/test_neuronexusio.py:17]: Yep. I got it in now that I have gin working on my laptop thanks to you!

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:27]: This was a note for myself. I don't know what the naming convention is for neuronexus (like what does pri actually stand for?). So what do we want to expose to the end user. Their `pri` names or their `ai0` names. We could do either?

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:251]: I think the parsing is super fast, but one thing I was thinking of was we could allow think about parsing more of header in this function then in spikeinterface or in neuroconv stuff you could run this function separately if you needed separate metadata then we could add a check in the main body that if this is already run we don't need to re-run. But the thought was inchoate (because honestly this function is so small right now). Maybe next time we have a call we can brainstorm on this because I don't think I'm explaining myself super well.

As far as naming, I just didn't want to use the same name so I added the fname tag onto it. But happy to change if it helps in conceptualizing more.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:159]: I need to check with @samuelgarcia, I honestly can't remember if he had any typing rules for the streams or not. I thought both indexes and ids were ints for the streams, but maybe that was just me seeing a pattern was there wasn't one.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:69]: Thanks.!

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:50]: Okay that's fair. I'll fix it in all places tomorrow !

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:55]: Any idea why? Seems wrong to me : P

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:126]: So you tried the different types until it worked?

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:149]: I mean showing/displaying the two timestamps in the error assertion.

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:159]: There are definitley some stream_ids that are strings. Plexon that I was just working on for example.

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:27]: Ah, got you, I thought you wer talking about slicing across channels.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:159]: Works for me then. Will change when I change names.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:55]: No that decision was before my time. Maybe at some point it was useful to initialize the class without an actual file for something in the testing suite? But honestly not sure. I thought it was strange to, but I just live with it now. :)

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:126]: Yes.... And I looked through comments in some thread where someone thought it was float32. So I iterated a little bit and found things that looked correct. For this to have no gain or offset it basically had to be some type of float.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:149]: That's what I thought you meant but wanted to confirm. Yep I will add that.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:27]: Sorry yeah the plexon issue with unclear channel naming was on my mind. So since we aren't do sub-streaming yet both the analog and digital channels all come out in the same stream (or buffer since I like the buffer term better). So the end-user needs to understand the channel naming scheme if they want to slice out the digital data and only look at the amplifier data.

Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:27]: Ah, perfect, that clarifies it all. So in the new terminology you are providing buffer based streams and once someone with more expertise on the format comes along then we can improve it to use "logic" based streams. Right?

File: neo/rawio/neuronexusrawio.py
Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py:55]: Yes, not that you mentioned it, I think at some point I removed one and it broke down stuff in the part of neo that I don't use : O (the not raw part)

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:27]: Yeah basically. I could technically figure it out, but since this is a buffer I figured I would leave as is for now to see what Sam was thinking of for the stream api to minimize the surface area of fixes. If a user complains then I could just implement something myself.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: Done.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: Done

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:104]: Just left for now.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: Fixed.

**Comment by zm711** [comment]
It would be relatively easy. My only concern is that if we have a universal way to do this then we are adding more work to adapt all of these rawios into the new format. But in this case the last six channels are reserved for the non-ephys data. So I could split these off intro logical streams.

**Comment by zm711** [comment]
Or we could merge this as is and then iterate for the logical streams in a new PR. That way if the buffer -> stream api happens before we have time then we fold that in and if this becomes an issue for users we can quickly fix it upon request before the api is added.

**Comment by zm711** [comment]
@samuelgarcia, this one can also take a review in your queue. Just pinging you now that you're back here. We have a couple requests on spikeinterface (and on neuroconv) for this. At your leisure of course.

**Comment by zm711** [comment]
C'est le ping à Sam de la semaine.

**Comment by h-mayorquin** [comment]
ceci n'est pas une ping

**Comment by zm711** [comment]
He's on the thread. He'll see it :P

File: neo/rawio/neuronexusrawio.py
Comment (samuelgarcia) [inline neo/rawio/neuronexusrawio.py:219]: Any chance to have a rec_datetime in the metadata dict ?

File: neo/rawio/neuronexusrawio.py
Comment (samuelgarcia) [inline neo/rawio/neuronexusrawio.py:237]: If I remember correctly, double slicing sometimes do not play so well when different kind of slicing

```suggestion
        raw_data = self._raw_data[i_start:i_stop, :]
        if channel_indexes is None:
            raw_data = raw_data[:, channel_indexes]
        return raw_data
```

**Comment by samuelgarcia** [comment]
Salut Zach.
This looks good to me.
I made one or two comments.
Sorry for the long delay.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:219]: I don't remember. Let me check!

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:237]: Works for me. I didn't know double-slicing could be bad (I think we do it in other places), so happy to split it!

Comment (samuelgarcia) [reply neo/rawio/neuronexusrawio.py]: It is a bit fuzzy but some stuff like mixing slice and bool mask in fancy indexing can be strange. Maybe it is not anymore the case. But maybe we have this in many places in neo. SO no worry

Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: I'm already fixing it :) We were missing one thing.

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [reply neo/rawio/neuronexusrawio.py:219]: Okay so I'm adding to the header like this:

```
self.header['rec_datetime'] = self.metadata['status']['start_time']
```
It is setup like: date and time. Do you want any other info?

**Comment by zm711** [comment]
Thanks Sam. I think this is ready to go now after your feedback!

File: neo/rawio/neuronexusrawio.py
Comment (samuelgarcia) [inline neo/rawio/neuronexusrawio.py:237]:
```suggestion
        if channel_indexes is not None:
```

Comment (samuelgarcia) [reply neo/rawio/neuronexusrawio.py:237]: My bad. Sorry.

**Review by samuelgarcia — APPROVED** [review]

File: neo/rawio/neuronexusrawio.py
Comment (samuelgarcia) [inline neo/rawio/neuronexusrawio.py:221]: @zm711 could check this instead ?
the rec_datetime must be a real datatime and this must routed inside raw_annotaion at block and segment level for instance like this

```suggestion
        rec_datetime = datetime.datetime(self.metadata['status']['start_time'])
        bl_annotations = self.raw_annotations["blocks"][0]
        seg_annotations = bl_annotations["segments"][0]
        for d in (bl_annotations, seg_annotations):
            d["rec_datetime"] = rec_datetime
```

File: neo/rawio/neuronexusrawio.py
Comment (zm711) [inline neo/rawio/neuronexusrawio.py:241]: @samuelgarcia so I don't do as much date time stuff you want to look at this. Basically we get this string. I don't know what "T" inside the string or the 4:00 at the end (maybe duration in seconds?) so I did some ugly parsing. If you have a better idea from the parsing let me know!

Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: I guess my fear with this type of parsing is that if they change how they encode it we end up with giant parsing logic.

Comment (h-mayorquin) [reply neo/rawio/neuronexusrawio.py]: 2024-07-01T13:04:49.4972245-04:00
date_time_{T}_{timestamps}_{time_zone}

Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: @h-mayorquin if you have a good system could you actually make a PR off of this one and I'll merge it into this. I'm trash at this parsing and it seems like you have a better idea of how to pull this out!

Comment (zm711) [reply neo/rawio/neuronexusrawio.py]: We are ready to merge this @h-mayorquin  unless you have a better idea for parsing. Do you want me to hold off until you take a look or good to merge and we can prettify this bit later?

**Review by samuelgarcia — APPROVED** [review]

**Comment by samuelgarcia** [comment]
merci

**Pull request merged**
